New sysctl filename - 99-protect-links.conf

gardouille opened the pull request:
Bullseye and before: /usr/lib/sysctl.d/protect-links.conf
Since Bookworm: /usr/lib/sysctl.d/99-protect-links.conf

gardouille committed:
New sysctl filename - 99-protect-links.conf

Bullseye and before: /usr/lib/sysctl.d/protect-links.conf
Since Bookworm: /usr/lib/sysctl.d/99-protect-links.conf

gardouille commented:
With previous filename:
```
ls -l /etc/sysctl.d
-rw-r--r-- 1 root root  653 Jun 26 14:38 10-memory.conf
-rw-r--r-- 1 root root 1493 Jun 26 14:38 20-network.conf
-rw-r--r-- 1 root root  715 Jun 26 14:38 50-pid-max.conf
-rw-r--r-- 1 root root  459 Jun 26 14:38 protect-links.conf
lrwxrwxrwx 1 root root   14 Feb 28 12:15 99-sysctl.conf -> ../sysctl.conf
-rw-r--r-- 1 root root  798 Dec 19  2022 README.sysctl
```

`sysctl` won't work:
```
sysctl --system
* Applying /etc/sysctl.d/10-memory.conf ...
* Applying /etc/sysctl.d/20-network.conf ...
* Applying /etc/sysctl.d/50-pid-max.conf ...
* Applying /usr/lib/sysctl.d/99-protect-links.conf ...
* Applying /etc/sysctl.d/99-sysctl.conf ...
* Applying /etc/sysctl.d/protect-links.conf ...
* Applying /etc/sysctl.conf ...
net.ipv4.ip_forward = 0
net.ipv6.conf.all.forwarding = 0
net.ipv6.conf.all.accept_ra = 0
net.ipv6.conf.default.accept_ra = 0
net.ipv4.conf.all.rp_filter = 1
net.ipv4.conf.default.rp_filter = 1
net.ipv4.icmp_echo_ignore_broadcasts = 1
net.ipv4.icmp_ignore_bogus_error_responses = 1
net.ipv4.icmp_ratelimit = 100
net.ipv4.icmp_ratemask = 88089
net.ipv4.tcp_timestamps = 0
net.ipv4.tcp_rfc1337 = 1
net.ipv4.tcp_slow_start_after_idle = 1
sysctl: permission denied on key "fs.protected_fifos"
sysctl: permission denied on key "fs.protected_hardlinks"
sysctl: permission denied on key "fs.protected_regular"
sysctl: permission denied on key "fs.protected_symlinks"
```

With modification, everything works fine:
```
sysctl --system
* Applying /etc/sysctl.d/10-memory.conf ...
* Applying /etc/sysctl.d/20-network.conf ...
* Applying /etc/sysctl.d/50-pid-max.conf ...
* Applying /etc/sysctl.d/99-protect-links.conf ...
* Applying /etc/sysctl.d/99-sysctl.conf ...
* Applying /etc/sysctl.d/protect-links.conf ...
* Applying /etc/sysctl.conf ...
net.ipv4.ip_forward = 0
…
net.ipv4.tcp_slow_start_after_idle = 1
```

And it doesn't change filename on a Bullseye host :+1:

prahal commented:
> ```
> sysctl --system
> (...)
> sysctl: permission denied on key "fs.protected_fifos"
> sysctl: permission denied on key "fs.protected_hardlinks"
> sysctl: permission denied on key "fs.protected_regular"
> sysctl: permission denied on key "fs.protected_symlinks"
> ```

The fix is good but does not fix the error (in both cases, ie the debops sysctl protect-links.conf values take precedence.
I only get this error when running `sysctl --system` in an lxc container.
It prevents me from completing the debops run on this lxc container even if I leave it at bullseye (not upgraded from bookworm).
On a bare metal bookworm host I get no error, even with the broken file name:
```
sudo LC_ALL=C sysctl --system
* Applying /etc/sysctl.d/0-sysrq.conf ...
* Applying /etc/sysctl.d/10-debops_sysctl.conf ...
* Applying /etc/sysctl.d/10-memory.conf ...
* Applying /etc/sysctl.d/10-ptrace.conf ...
* Applying /etc/sysctl.d/20-network.conf ...
* Applying /etc/sysctl.d/30-ferm.conf ...
* Applying /etc/sysctl.d/30-lxc-inotify.conf ...
* Applying /etc/sysctl.d/30-postgresql-shm.conf ...
* Applying /etc/sysctl.d/30-spideroakone.conf ...
* Applying /usr/lib/sysctl.d/30-tracker.conf ...
* Applying /usr/lib/sysctl.d/50-bubblewrap.conf ...
* Applying /usr/lib/sysctl.d/50-coredump.conf ...
* Applying /etc/sysctl.d/50-pid-max.conf ...
* Applying /usr/lib/sysctl.d/99-protect-links.conf ...
* Applying /etc/sysctl.d/99-sysctl.conf ...
* Applying /etc/sysctl.d/protect-links.conf ...
* Applying /etc/sysctl.conf ...
kernel.sysrq = 1
kernel.shmall = 1396530
kernel.shmmax = 3575115776
net.ipv4.conf.all.rp_filter = 1
net.ipv4.conf.default.rp_filter = 1
net.ipv4.icmp_echo_ignore_broadcasts = 1
net.ipv4.icmp_ignore_bogus_error_responses = 1
net.ipv4.icmp_ratelimit = 100
net.ipv4.icmp_ratemask = 88089
net.ipv4.ip_forward = 0
net.ipv4.tcp_rfc1337 = 1
net.ipv4.tcp_timestamps = 0
net.ipv6.conf.all.accept_ra = 0
net.ipv6.conf.all.forwarding = 0
net.ipv6.conf.default.accept_ra = 0
vm.swappiness = 60
vm.vfs_cache_pressure = 100
kernel.shmmax = 4040683520
kernel.shmall = 1578392
vm.swappiness = 60
vm.vfs_cache_pressure = 100
kernel.yama.ptrace_scope = 0
net.ipv4.ip_forward = 0
net.ipv6.conf.all.forwarding = 0
net.ipv6.conf.all.accept_ra = 0
net.ipv6.conf.default.accept_ra = 0
net.ipv4.conf.all.rp_filter = 1
net.ipv4.conf.default.rp_filter = 1
net.ipv4.icmp_echo_ignore_broadcasts = 1
net.ipv4.icmp_ignore_bogus_error_responses = 1
net.ipv4.icmp_ratelimit = 100
net.ipv4.icmp_ratemask = 88089
net.ipv4.tcp_timestamps = 0
net.ipv4.tcp_rfc1337 = 1
net.ipv4.tcp_slow_start_after_idle = 1
fs.inotify.max_user_instances = 1024
fs.inotify.max_user_watches = 65536
fs.inotify.max_user_watches = 65536
kernel.unprivileged_userns_clone = 1
kernel.core_pattern = |/lib/systemd/systemd-coredump %P %u %g %s %t 9223372036854775808 %h
kernel.core_pipe_limit = 16
fs.suid_dumpable = 2
kernel.pid_max = 4194304
fs.protected_fifos = 1
fs.protected_hardlinks = 1
fs.protected_regular = 2
fs.protected_symlinks = 1
fs.protected_fifos = 1
fs.protected_hardlinks = 1
fs.protected_regular = 2
fs.protected_symlinks = 1
```

Could you confirm you can get the above error when not in an lxc container? And that your fix resolves the issue inside an lxc container (which it should not)?

It might be that the sysctl in a Debian  lxc container is already broken in Deian upstream because the error happens for keys that are in the Debian upstream diverted /etc/sysctl.d/protect-links.conf.dpkg-divert in bullseye.

gardouille commented:
Yes, this issue is specific to LXC hosts, as mentionned in [parameter comment](https://github.com/debops/debops/blob/master/ansible/roles/sysctl/defaults/main.yml#L263). ;)
On LXC hosts, the file will contain commented parameters to avoid such problem.

And yes, this commit solves sysctl problem with new Bookworm installation only.

For Bookworm hosts where DebOps has already been applied, it will be necessary to manually run `rm -f /etc/sysctl.d/protect-links.conf`, especially to solve the problème on LXC hosts ;)
Or… I can add a condition to the "protect-links" parameter in sysctl__default_parameters (debian≤11) and duplicate it with a debian≥12 condition. But two "similar" parameters and Bookworm is not yet supported…

@drybjed Any preference ?

drybjed commented:
I think that it's better to keep things simple. I'll merge this PR and new DebOps installations will not have this problem. We know the fix for the current installations - remove `/etc/sysctl.d/protect-links.conf` and it is fixed.

drybjed reviewed (approved):
Thanks for the fix! I'll improve the formatting a bit, and backport this to the stable releaes.

prahal commented:
@gardouille 

> Yes, this issue is specific to LXC hosts, as mentionned in [parameter comment](https://github.com/debops/debops/blob/master/ansible/roles/sysctl/defaults/main.yml#L263). ;) On LXC hosts, the file will contain commented parameters to avoid such problem.
> 
> And yes, this commit solves sysctl problem with new Bookworm installation only.
> 
> For Bookworm hosts where DebOps has already been applied, it will be necessary to manually run `rm -f /etc/sysctl.d/protect-links.conf`, especially to solve the problème on LXC hosts ;) Or… I can add a condition to the "protect-links" parameter in sysctl__default_parameters (debian≤11) and duplicate it with a debian≥12 condition. But two "similar" parameters and Bookworm is not yet supported…

I guess you mean Bullseye instead of Bookworm (for those not handled by your fix) and lxc guest, not host?
I agree that the fact that '/proc/sys/fs/' namespace is usually read-only in unprivileged LXC is mentioned in https://github.com/debops/debops/blob/master/ansible/roles/sysctl/defaults/main.yml#L263 but where are those turned off for LXC guests (be it for Bookworm or Bullseye)?
Are you sure you fixed the issue for bookworm LXC guests? 
Mind the fix is correct but make no functional changes as it keeps the same values as the file it diverts.

I removed the /etc/sysctl.d/protect-links.conf on the bullseye LX guest and ran debops against it anew and it fails on the fs sysctl parameters.

This might deserve its own bug report though. But could you confirm you can run `sudo sysctl --system` in your lxc bookworm guest and have no error?
The cat is debops only attempts to run `sysctl --system` when the sysctl.d file is changed. So if you make a second debops run then no error triggers as sysctl is not run at all (this is another debops bug that if a run fails to apply settings the next run behaves as if the settings are already applied and do not apply them at all).

gardouille commented:
Yes, LXC guests, my bad ;)

Double-checking on a new LXC guest (from Proxmox Debian 12 template fyi).
<details>
<summary><b>New Bookworm "empty" guest</b></summary>

```
root@testme:~# cat /etc/debian_version
12.0

root@testme:~# ls -l /etc/sysctl.d
total 4
lrwxrwxrwx 1 root root  14 Feb 28 12:15 99-sysctl.conf -> ../sysctl.conf
-rw-r--r-- 1 root root 798 Dec 19  2022 README.sysctl
```

And, as expected, the already installed `sysctl` fails (default config):

```
root@testme:~# sysctl --system
* Applying /usr/lib/sysctl.d/50-pid-max.conf ...
* Applying /usr/lib/sysctl.d/99-protect-links.conf ...
* Applying /etc/sysctl.d/99-sysctl.conf ...
* Applying /etc/sysctl.conf ...
sysctl: permission denied on key "kernel.pid_max"
sysctl: permission denied on key "fs.protected_fifos"
sysctl: permission denied on key "fs.protected_hardlinks"
sysctl: permission denied on key "fs.protected_regular"
sysctl: permission denied on key "fs.protected_symlinks"
```
</details>

<details>
<summary><b>Controller with protect-links.conf filename</b></summary>
roles/sysctl/defaults/main.yml

```
…
  - name: 'protect-links'
    filename: 'protect-links.conf'
    divert: True
    comment: |
      Protected links

      Protects against creating or following links under certain conditions
…
```
</details>

<details>
<summary><b>Controller: bootstrap-ldap 👍</b></summary>

```
PLAY RECAP *******************************************************************************************************
testme                     : ok=239  changed=115  unreachable=0    failed=0    skipped=122  rescued=0    ignored=0
```
</details>

<details>
<summary><b>Controller: debops run site fails 👎</b></summary>

```
…
TASK [debops.debops.sysctl : Apply kernel parameters if they were modified] *****************************************                                                                 [12/4144]
fatal: [testme]: FAILED! => changed=true
  cmd: |-
    set -o nounset -o pipefail -o errexit &&
    sysctl --system
  delta: '0:00:00.021784'
  end: '2023-06-28 17:26:38.518693'
  msg: non-zero return code
  rc: 1
  start: '2023-06-28 17:26:38.496909'
  stderr: |-
    sysctl: permission denied on key "fs.protected_fifos"
    sysctl: permission denied on key "fs.protected_hardlinks"
    sysctl: permission denied on key "fs.protected_regular"
    sysctl: permission denied on key "fs.protected_symlinks"
  stderr_lines: <omitted>
  stdout: |-
    * Applying /etc/sysctl.d/10-memory.conf ...
    * Applying /etc/sysctl.d/20-network.conf ...
    * Applying /etc/sysctl.d/50-pid-max.conf ...
    * Applying /usr/lib/sysctl.d/99-protect-links.conf ...
    * Applying /etc/sysctl.d/99-sysctl.conf ...
    * Applying /etc/sysctl.d/protect-links.conf ...
    * Applying /etc/sysctl.conf ...
    net.ipv4.ip_forward = 0
    net.ipv6.conf.all.forwarding = 0
    net.ipv6.conf.all.accept_ra = 0
    net.ipv6.conf.default.accept_ra = 0
    net.ipv4.conf.all.rp_filter = 1
    net.ipv4.conf.default.rp_filter = 1
    net.ipv4.icmp_echo_ignore_broadcasts = 1
    net.ipv4.icmp_ignore_bogus_error_responses = 1
    net.ipv4.icmp_ratelimit = 100
    net.ipv4.icmp_ratemask = 88089
    net.ipv4.tcp_timestamps = 0
    net.ipv4.tcp_rfc1337 = 1
    net.ipv4.tcp_slow_start_after_idle = 1
  stdout_lines: <omitted>

PLAY RECAP **********************************************************************************************************
testme                     : ok=357  changed=53   unreachable=0    failed=1    skipped=240  rescued=0    ignored=0
```
</details>

<details>
<summary><b>Controller: update filename to 99-protect-links.conf (this fix)</b></summary>
roles/sysctl/defaults/main.yml

```
…
  - name: 'protect-links'
    filename: '{{ "protect-links.conf"
        if (ansible_distribution_release in ["stretch", "buster", "bullseye"])
        else "99-protect-links.conf" }}'
    divert: True
```
</details>

<details>
<summary><b>Controller: debops run service sysctl succeed 👍</b></summary>

```
Executing Ansible playbooks:                                                                                                                                                          [11/4202]
~/src/debops3-venv/lib/python3.11/site-packages/debops/_data/ansible/collections/ansible_collections/debops/debops/playbooks/service/sysctl.yml

PLAY [Manage kernel parameters using sysctl] **************************************************************************************************************************************************

TASK [Gathering Facts] ************************************************************************************************************************************************************************
ok: [testme]

TASK [debops.debops.sysctl : Make sure that Ansible local facts directory exists] *************************************************************************************************************
ok: [testme]

TASK [debops.debops.sysctl : Save sysctl local facts] *****************************************************************************************************************************************
ok: [testme]

TASK [debops.debops.sysctl : Add/remove diversion of custom sysctl configuration files] *******************************************************************************************************
changed: [testme] => (item={'name': 'protect-links', 'state': 'present'})

TASK [debops.debops.sysctl : Generate custom sysctl configuration files] **********************************************************************************************************************
ok: [testme] => (item=memory)
changed: [testme] => (item=protect-links)
ok: [testme] => (item=network)
ok: [testme] => (item=pid-max)

TASK [debops.debops.sysctl : Check sysctl command capabilities] *******************************************************************************************************************************
changed: [testme]

TASK [debops.debops.sysctl : Apply kernel parameters if they were modified] *******************************************************************************************************************
changed: [testme]

TASK [debops.debops.sysctl : Post hooks] ******************************************************************************************************************************************************

PLAY RECAP ************************************************************************************************************************************************************************************
testme                     : ok=7    changed=4    unreachable=0    failed=0    skipped=1    rescued=0    ignored=0
```
</details>

<details>
<summary><b>LXC guest: sysctl succeed</b></summary>
Current state:

```
root@testme:~# ls -l /etc/sysctl.d
total 24
-rw-r--r-- 1 root root  653 Jun 28 17:26 10-memory.conf
-rw-r--r-- 1 root root 1493 Jun 28 17:26 20-network.conf
-rw-r--r-- 1 root root  715 Jun 28 17:26 50-pid-max.conf
-rw-r--r-- 1 root root  459 Jun 28 17:30 99-protect-links.conf              <======= New for Debian 12
lrwxrwxrwx 1 root root   14 Feb 28 12:15 99-sysctl.conf -> ../sysctl.conf
-rw-r--r-- 1 root root  459 Jun 28 17:26 protect-links.conf                 <======= "Old" for Debian ≤ 11
-rw-r--r-- 1 root root  798 Dec 19  2022 README.sysctl
```

99-protect-links.conf content :

```
# This file is managed remotely, all changes will be lost

# Protected links
#
# Protects against creating or following links under certain conditions
# Debian kernels have both set to 1 (restricted)
# See https://www.kernel.org/doc/Documentation/sysctl/fs.txt

# Read-only parameter.
#fs.protected_fifos = 1
# Read-only parameter.
#fs.protected_hardlinks = 1
# Read-only parameter.
#fs.protected_regular = 2
# Read-only parameter.
#fs.protected_symlinks = 1
```

`sysctl` seems ok (no matter how many tries) :

```
root@testme:~# sysctl --system
* Applying /etc/sysctl.d/10-memory.conf ...
* Applying /etc/sysctl.d/20-network.conf ...
* Applying /etc/sysctl.d/50-pid-max.conf ...
* Applying /etc/sysctl.d/99-protect-links.conf ...
* Applying /etc/sysctl.d/99-sysctl.conf ...
* Applying /etc/sysctl.d/protect-links.conf ...
* Applying /etc/sysctl.conf ...
net.ipv4.ip_forward = 0
net.ipv6.conf.all.forwarding = 0
net.ipv6.conf.all.accept_ra = 0
net.ipv6.conf.default.accept_ra = 0
net.ipv4.conf.all.rp_filter = 1
net.ipv4.conf.default.rp_filter = 1
net.ipv4.icmp_echo_ignore_broadcasts = 1
net.ipv4.icmp_ignore_bogus_error_responses = 1
net.ipv4.icmp_ratelimit = 100
net.ipv4.icmp_ratemask = 88089
net.ipv4.tcp_timestamps = 0
net.ipv4.tcp_rfc1337 = 1
net.ipv4.tcp_slow_start_after_idle = 1
```
</details>

<details>
<summary><b>LXC guest: Clean old file to avoid multiple definition of a parameter</b></summary>
Remove old file

```
root@testme:~# rm -f /etc/sysctl.d/protect-links.conf

root@testme:~# ls -l /etc/sysctl.d
total 24
-rw-r--r-- 1 root root  653 Jun 28 17:26 10-memory.conf
-rw-r--r-- 1 root root 1493 Jun 28 17:26 20-network.conf
-rw-r--r-- 1 root root  715 Jun 28 17:26 50-pid-max.conf
-rw-r--r-- 1 root root  459 Jun 28 17:30 99-protect-links.conf              <======= New
lrwxrwxrwx 1 root root   14 Feb 28 12:15 99-sysctl.conf -> ../sysctl.conf
-rw-r--r-- 1 root root  798 Dec 19  2022 README.sysctl
```

`sysctl` is still fine (no matter how many tries) :
```
root@testme:~# sysctl --system
* Applying /etc/sysctl.d/10-memory.conf ...
* Applying /etc/sysctl.d/20-network.conf ...
* Applying /etc/sysctl.d/50-pid-max.conf ...
* Applying /etc/sysctl.d/99-protect-links.conf ...
* Applying /etc/sysctl.d/99-sysctl.conf ...
* Applying /etc/sysctl.d/protect-links.conf ...
* Applying /etc/sysctl.conf ...
net.ipv4.ip_forward = 0
net.ipv6.conf.all.forwarding = 0
net.ipv6.conf.all.accept_ra = 0
net.ipv6.conf.default.accept_ra = 0
net.ipv4.conf.all.rp_filter = 1
net.ipv4.conf.default.rp_filter = 1
net.ipv4.icmp_echo_ignore_broadcasts = 1
net.ipv4.icmp_ignore_bogus_error_responses = 1
net.ipv4.icmp_ratelimit = 100
net.ipv4.icmp_ratemask = 88089
net.ipv4.tcp_timestamps = 0
net.ipv4.tcp_rfc1337 = 1
net.ipv4.tcp_slow_start_after_idle = 1
```
</details>

<details>
<summary><b>Controller: debops run service sysctl has nothing to do 👍</b></summary>

```
…
TASK [debops.debops.sysctl : Post hooks] ******************************************************************************************************************************************************

PLAY RECAP ************************************************************************************************************************************************************************************
testme                     : ok=5    changed=0    unreachable=0    failed=0    skipped=3    rescued=0    ignored=0
…
```
</details>


Does this help ? :crossed_fingers:

gardouille commented:
And, an LXC Bullseye guest should be fine with **/etc/sysctl.d/protect-links.conf** file :+1:

prahal commented:
Thank you @gardouille . My protected-links.conf fs lines are not commented out.

It turned out that my privileged container config from back when I migrated to LXC 3 was either not up to date or plain incorrect. I created a new privileged container following https://docs.debops.org/en/v3.0.0/ansible/roles/lxc/getting-started.html and compared it with mine. Fixing the difference and all is fine now (the container is still bullseye for now).
I get `/etc/ansible/facts.d/sysctl.fact` without fs in the writable section (only net, in fact, is there), and now `service/sysctl` properly generates the fs lines commented in `/etc/sysctl.d/protected-links.conf` for the container.
